File: src/services/DatasetManager.ts
/**
 * Dataset Manager for Diverse and Representative Training Data
 * Ensures balanced representation across all demographic groups
 */

import {
  DemographicGroup,
  DemographicStats,
  RetinaImageWithDemographics,
  DatasetInfo,
} from '../types/fairness';

export class DatasetManager {
  private datasets = new Map<string, DatasetInfo>();
  private demographicStats = new Map<string, DemographicStats>();
  private readonly MIN_REPRESENTATION = 0.05; // 5% minimum representation per group

  /**
   * Load diverse datasets from multiple global sources
   */
  async loadDiverseDatasets(): Promise<void> {
    const datasets: Array<{
      name: string;
      source: string;
      demographics: DemographicGroup;
    }> = [
      {
        name: 'APTOS',
        source: 'India',
        demographics: {
          race: ['South Asian'],
          ethnicity: ['Indian'],
          gender: ['male', 'female'],
          ageRanges: ['20-30', '30-40', '40-50', '50-60', '60+'],
          geographicRegions: ['South Asia'],
          socioeconomicStatus: ['low', 'middle', 'high'],
        },
      },
      {
        name: 'EyePACS',
        source: 'USA',
        demographics: {
          race: ['White', 'Black', 'Asian', 'Hispanic', 'Native American'],
          ethnicity: ['Non-Hispanic', 'Hispanic'],
          gender: ['male', 'female', 'other'],
          ageRanges: ['20-30', '30-40', '40-50', '50-60', '60+'],
          geographicRegions: ['North America'],
          socioeconomicStatus: ['low', 'middle', 'high'],
        },
      },
      {
        name: 'Messidor',
        source: 'France',
        demographics: {
          race: ['White', 'North African', 'Sub-Saharan African'],
          ethnicity: ['European', 'African'],
          gender: ['male', 'female'],
          ageRanges: ['40-50', '50-60', '60+'],
          geographicRegions: ['Europe'],
          socioeconomicStatus: ['middle', 'high'],
        },
      },
      {
        name: 'RFMiD',
        source: 'Multi-country',
        demographics: {
          race: ['South Asian', 'Southeast Asian', 'Middle Eastern'],
          ethnicity: ['Various'],
          gender: ['male', 'female'],
          ageRanges: ['20-30', '30-40', '40-50', '50-60', '60+'],
          geographicRegions: ['South Asia', 'Southeast Asia', 'Middle East'],
          socioeconomicStatus: ['low', 'middle', 'high'],
        },
      },
    ];

    for (const dataset of datasets) {
      await this.loadDataset(dataset);
    }
  }

  /**
   * Load and validate a dataset
   */
  private async loadDataset(dataset: {
    name: string;
    source: string;
    demographics: DemographicGroup;
  }): Promise<void> {
    console.log(`Loading ${dataset.name} from ${dataset.source}`);

    // In a real implementation, this would load actual images
    // For now, we simulate dataset loading
    const mockStats = this.calculateDemographicStats(dataset);
    this.demographicStats.set(dataset.name, mockStats);

    // Check for representation gaps
    this.identifyRepresentationGaps(mockStats, dataset.demographics);

    const datasetInfo: DatasetInfo = {
      name: dataset.name,
      source: dataset.source,
      demographics: dataset.demographics,
      totalSamples: mockStats.totalSamples,
      representation: {
        ...mockStats.raceDistribution,
        ...mockStats.ethnicityDistribution,
        ...mockStats.genderDistribution,
      },
    };

    this.datasets.set(dataset.name, datasetInfo);
  }

  /**
   * Calculate demographic statistics for a dataset
   */
  calculateDemographicStats(dataset: {
    demographics: DemographicGroup;
  }): DemographicStats {
    // In a real implementation, this would analyze actual image data
    // For now, we return mock statistics
    const stats: DemographicStats = {
      totalSamples: 1000, // Mock value
      raceDistribution: {},
      ethnicityDistribution: {},
      genderDistribution: {},
      ageDistribution: {},
      geographicDistribution: {},
    };

    // Simulate distribution calculations
    dataset.demographics.race.forEach((race) => {
      stats.raceDistribution[race] = Math.random() * 0.3 + 0.1;
    });

    dataset.demographics.gender.forEach((gender) => {
      stats.genderDistribution[gender] = Math.random() * 0.4 + 0.2;
    });

    return stats;
  }

  /**
   * Identify representation gaps in dataset
   */
  identifyRepresentationGaps(
    stats: DemographicStats,
    targetDemographics: DemographicGroup
  ): void {
    const gaps: Array<{
      demographic: string;
      group: string;
      representation: number;
      minimumRequired: number;
    }> = [];

    // Check race representation
    targetDemographics.race.forEach((race) => {
      const representation = stats.raceDistribution[race] || 0;
      if (representation < this.MIN_REPRESENTATION) {
        gaps.push({
          demographic: 'race',
          group: race,
          representation,
          minimumRequired: this.MIN_REPRESENTATION,
        });
      }
    });

    // Check gender representation
    targetDemographics.gender.forEach((gender) => {
      const representation = stats.genderDistribution[gender] || 0;
      if (representation < this.MIN_REPRESENTATION) {
        gaps.push({
          demographic: 'gender',
          group: gender,
          representation,
          minimumRequired: this.MIN_REPRESENTATION,
        });
      }
    });

    if (gaps.length > 0) {
      console.warn('Representation gaps identified:', gaps);
      this.triggerDataCollection(gaps);
    }
  }

  /**
   * Ensure balanced training split across demographics
   */
  async ensureBalancedTrainingSplit(
    images: RetinaImageWithDemographics[]
  ): Promise<RetinaImageWithDemographics[]> {
    const grouped = this.groupByDemographics(images);
    const balancedSplit: RetinaImageWithDemographics[] = [];

    // Ensure minimum representation from each demographic group
    const numGroups = Object.keys(grouped).length;
    const minPerGroup = Math.floor((images.length / numGroups) * 0.1);

    for (const [_group, groupImages] of Object.entries(grouped)) {
      const samples = this.sampleBalanced(groupImages, minPerGroup);
      balancedSplit.push(...samples);
    }

    // Fill remaining slots with stratified random sampling
    const remaining = images.length - balancedSplit.length;
    if (remaining > 0) {
      const remainingSamples = this.stratifiedRandomSample(
        images,
        balancedSplit,
        remaining
      );
      balancedSplit.push(...remainingSamples);
    }

    return balancedSplit;
  }

  /**
   * Group images by demographic attributes
   */
  private groupByDemographics(
    images: RetinaImageWithDemographics[]
  ): Record<string, RetinaImageWithDemographics[]> {
    const groups: Record<string, RetinaImageWithDemographics[]> = {};

    images.forEach((image) => {
      const key = this.getDemographicGroupKey(image);
      if (!groups[key]) {
        groups[key] = [];
      }
      groups[key].push(image);
    });

    return groups;
  }

  /**
   * Generate a key for demographic grouping
   */
  private getDemographicGroupKey(image: RetinaImageWithDemographics): string {
    const demo = image.demographics;
    const ageRange = demo.age ? this.getAgeRange(demo.age) : 'unknown';
    return `${demo.race || 'unknown'}-${demo.ethnicity || 'unknown'}-${
      demo.gender || 'unknown'
    }-${ageRange}`;
  }

  /**
   * Get age range from age
   */
  private getAgeRange(age: number): string {
    if (age < 30) return '20-30';
    if (age < 40) return '30-40';
    if (age < 50) return '40-50';
    if (age < 60) return '50-60';
    return '60+';
  }

  /**
   * Sample balanced images from a group
   */
  private sampleBalanced(
    images: RetinaImageWithDemographics[],
    minCount: number
  ): RetinaImageWithDemographics[] {
    const shuffled = [...images].sort(() => Math.random() - 0.5);
    return shuffled.slice(0, Math.min(minCount, images.length));
  }

  /**
   * Stratified random sampling
   */
  private stratifiedRandomSample(
    allImages: RetinaImageWithDemographics[],
    alreadySelected: RetinaImageWithDemographics[],
    count: number
  ): RetinaImageWithDemographics[] {
    const selectedIds = new Set(alreadySelected.map((img) => img.id));
    const available = allImages.filter((img) => !selectedIds.has(img.id));
    const shuffled = [...available].sort(() => Math.random() - 0.5);
    return shuffled.slice(0, count);
  }

  /**
   * Trigger data collection for underrepresented groups
   */
  private triggerDataCollection(gaps: Array<{
    demographic: string;
    group: string;
    representation: number;
    minimumRequired: number;
  }>): void {
    console.log('Triggering data collection for gaps:', gaps);
    // In production, this would trigger actual data collection workflows
  }

  /**
   * Get all loaded datasets
   */
  getDatasets(): DatasetInfo[] {
    return Array.from(this.datasets.values());
  }

  /**
   * Get demographic statistics for a dataset
   */
  getDemographicStats(datasetName: string): DemographicStats | undefined {
    return this.demographicStats.get(datasetName);
  }
}

File: services/dataset_manager.py
"""
Dataset Manager for Diverse and Representative Training Data
Ensures balanced representation across all demographic groups
"""

import numpy as np
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from collections import defaultdict
import logging

logger = logging.getLogger(__name__)


@dataclass
class DemographicInfo:
    """Demographic information for a sample"""
    race: Optional[str] = None
    ethnicity: Optional[str] = None
    gender: Optional[str] = None
    age: Optional[int] = None
    geographic_region: Optional[str] = None
    socioeconomic_status: Optional[str] = None


class DatasetManager:
    """
    Manages diverse datasets and ensures balanced representation
    """
    
    def __init__(self, min_representation: float = 0.05):
        """
        Initialize dataset manager
        
        Args:
            min_representation: Minimum representation fraction (5% default)
        """
        self.min_representation = min_representation
        self.datasets = {}
        self.demographic_stats = {}
    
    def load_diverse_datasets(self) -> Dict[str, Dict]:
        """
        Load diverse datasets from multiple global sources
        
        Returns:
            Dictionary of dataset information
        """
        datasets = {
            'APTOS': {
                'source': 'India',
                'demographics': {
                    'race': ['South Asian'],
                    'ethnicity': ['Indian'],
                    'gender': ['male', 'female'],
                    'age_ranges': ['20-30', '30-40', '40-50', '50-60', '60+'],
                    'geographic_regions': ['South Asia'],
                    'socioeconomic_status': ['low', 'middle', 'high'],
                }
            },
            'EyePACS': {
                'source': 'USA',
                'demographics': {
                    'race': ['White', 'Black', 'Asian', 'Hispanic', 'Native American'],
                    'ethnicity': ['Non-Hispanic', 'Hispanic'],
                    'gender': ['male', 'female', 'other'],
                    'age_ranges': ['20-30', '30-40', '40-50', '50-60', '60+'],
                    'geographic_regions': ['North America'],
                    'socioeconomic_status': ['low', 'middle', 'high'],
                }
            },
            'Messidor': {
                'source': 'France',
                'demographics': {
                    'race': ['White', 'North African', 'Sub-Saharan African'],
                    'ethnicity': ['European', 'African'],
                    'gender': ['male', 'female'],
                    'age_ranges': ['40-50', '50-60', '60+'],
                    'geographic_regions': ['Europe'],
                    'socioeconomic_status': ['middle', 'high'],
                }
            },
            'RFMiD': {
                'source': 'Multi-country',
                'demographics': {
                    'race': ['South Asian', 'Southeast Asian', 'Middle Eastern'],
                    'ethnicity': ['Various'],
                    'gender': ['male', 'female'],
                    'age_ranges': ['20-30', '30-40', '40-50', '50-60', '60+'],
                    'geographic_regions': ['South Asia', 'Southeast Asia', 'Middle East'],
                    'socioeconomic_status': ['low', 'middle', 'high'],
                }
            }
        }
        
        for name, info in datasets.items():
            self.datasets[name] = info
            logger.info(f"Loaded dataset: {name} from {info['source']}")
        
        return datasets
    
    def calculate_demographic_stats(
        self,
        images: List[Tuple],  # List of (image_data, demographic_info, label)
    ) -> Dict[str, Dict]:
        """
        Calculate demographic statistics for a dataset
        
        Args:
            images: List of (image_data, demographic_info, label) tuples
            
        Returns:
            Dictionary of demographic statistics
        """
        stats = {
            'total_samples': len(images),
            'race_distribution': defaultdict(int),
            'ethnicity_distribution': defaultdict(int),
            'gender_distribution': defaultdict(int),
            'age_distribution': defaultdict(int),
            'geographic_distribution': defaultdict(int),
            'socioeconomic_distribution': defaultdict(int),
        }
        
        for _, demo_info, _ in images:
            if demo_info.race:
                stats['race_distribution'][demo_info.race] += 1
            if demo_info.ethnicity:
                stats['ethnicity_distribution'][demo_info.ethnicity] += 1
            if demo_info.gender:
                stats['gender_distribution'][demo_info.gender] += 1
            if demo_info.age:
                age_range = self._get_age_range(demo_info.age)
                stats['age_distribution'][age_range] += 1
            if demo_info.geographic_region:
                stats['geographic_distribution'][demo_info.geographic_region] += 1
            if demo_info.socioeconomic_status:
                stats['socioeconomic_distribution'][demo_info.socioeconomic_status] += 1
        
        # Convert to fractions
        total = len(images)
        for key in ['race_distribution', 'ethnicity_distribution', 'gender_distribution',
                    'age_distribution', 'geographic_distribution', 'socioeconomic_distribution']:
            stats[key] = {
                k: v / total for k, v in stats[key].items()
            }
        
        return stats
    
    def identify_representation_gaps(
        self,
        stats: Dict[str, Dict],
        target_demographics: Dict
    ) -> List[Dict]:
        """
        Identify representation gaps in dataset
        
        Args:
            stats: Calculated demographic statistics
            target_demographics: Target demographic distribution
            
        Returns:
            List of identified gaps
        """
        gaps = []
        
        # Check race representation
        for race in target_demographics.get('race', []):
            representation = stats['race_distribution'].get(race, 0)
            if representation < self.min_representation:
                gaps.append({
                    'demographic': 'race',
                    'group': race,
                    'representation': representation,
                    'minimum_required': self.min_representation
                })
        
        # Check gender representation
        for gender in target_demographics.get('gender', []):
            representation = stats['gender_distribution'].get(gender, 0)
            if representation < self.min_representation:
                gaps.append({
                    'demographic': 'gender',
                    'group': gender,
                    'representation': representation,
                    'minimum_required': self.min_representation
                })
        
        if gaps:
            logger.warning(f"Representation gaps identified: {gaps}")
        
        return gaps
    
    def ensure_balanced_training_split(
        self,
        images: List[Tuple],
    ) -> List[Tuple]:
        """
        Ensure balanced training split across demographics
        
        Args:
            images: List of (image_data, demographic_info, label) tuples
            
        Returns:
            Balanced subset of images
        """
        grouped = self._group_by_demographics(images)
        balanced_split = []
        
        # Ensure minimum representation from each demographic group
        num_groups = len(grouped)
        min_per_group = max(1, int(len(images) / num_groups * 0.1))
        
        for group_images in grouped.values():
            samples = self._sample_balanced(group_images, min_per_group)
            balanced_split.extend(samples)
        
        # Fill remaining slots with stratified random sampling
        remaining = len(images) - len(balanced_split)
        if remaining > 0:
            selected_ids = {id(img) for img, _, _ in balanced_split}
            available = [(img, demo, label) for img, demo, label in images 
                        if id(img) not in selected_ids]
            remaining_samples = self._stratified_random_sample(
                available, remaining
            )
            balanced_split.extend(remaining_samples)
        
        return balanced_split
    
    def _group_by_demographics(
        self,
        images: List[Tuple]
    ) -> Dict[str, List[Tuple]]:
        """Group images by demographic attributes"""
        groups = defaultdict(list)
        
        for img, demo_info, label in images:
            key = self._get_demographic_group_key(demo_info)
            groups[key].append((img, demo_info, label))
        
        return dict(groups)
    
    def _get_demographic_group_key(self, demo_info: DemographicInfo) -> str:
        """Generate a key for demographic grouping"""
        age_range = self._get_age_range(demo_info.age) if demo_info.age else 'unknown'
        return f"{demo_info.race or 'unknown'}-{demo_info.ethnicity or 'unknown'}-" \
               f"{demo_info.gender or 'unknown'}-{age_range}"
    
    def _get_age_range(self, age: int) -> str:
        """Get age range from age"""
        if age < 30:
            return '20-30'
        elif age < 40:
            return '30-40'
        elif age < 50:
            return '40-50'
        elif age < 60:
            return '50-60'
        else:
            return '60+'
    
    def _sample_balanced(
        self,
        images: List[Tuple],
        min_count: int
    ) -> List[Tuple]:
        """Sample balanced images from a group"""
        np.random.shuffle(images)
        return images[:min(min_count, len(images))]
    
    def _stratified_random_sample(
        self,
        images: List[Tuple],
        count: int
    ) -> List[Tuple]:
        """Stratified random sampling"""
        np.random.shuffle(images)
        return images[:count]
